baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/IllegalSafeLoggingArgument.java

```diff
@@ -427,4 +462,43 @@ public Description matchClass(ClassTree tree, VisitorState state) {
                         "Dangerous type: annotated '%s' but ancestors declare '%s'.", directSafety, ancestorSafety))
                 .build();
     }
+
+    @Override
+    public Description matchLambdaExpression(LambdaExpressionTree tree, VisitorState state) {
+        Safety requiredReturnSafety = getLambdaRequiredReturnSafety(tree, state);
+
+        if (requiredReturnSafety.allowsAll()) {
+            // Short-circuit if the return type allows all values
+            return Description.NO_MATCH;
+        }
+
+        Safety resultSafety = Safety.UNKNOWN;
+        switch (tree.getBodyKind()) {
+            case EXPRESSION:
+                resultSafety = SafetyAnalysis.of(state.withPath(new TreePath(state.getPath(), tree.getBody())));
+                break;
+            case STATEMENT:
+                // Shortcut - statement lambdas get their return type checked in the return statement matcher
+                // This also allows us to indicate which return statement is bad (if any) rather than the lambda itself
+                return Description.NO_MATCH;
+        }
+
+        if (requiredReturnSafety.allowsValueWith(resultSafety)) {
+            return Description.NO_MATCH;
+        }
+
+        return buildDescription(tree)
+                .setMessage(String.format(
+                        "Dangerous return value: result is '%s' but the lambda expects return '%s'.",
+                        resultSafety, requiredReturnSafety))
+                .build();
+    }
+
+    private Safety getLambdaRequiredReturnSafety(LambdaExpressionTree tree, VisitorState state) {
+        TargetType returnType = ASTHelpers.targetType(state.withPath(new TreePath(state.getPath(), tree)));
```

Comment:
Ah, very nice! This ends up calling `state.getTypes().findDescriptorType(ASTHelpers.getType(lambdaExpressionTree)).getReturnType()`, where the `findDescriptorType` provides the functional interface method contract

Reply:
Does this catch issues with switch expressions as well?

Reply:
Ah I haven't checked. Do you have an example in mind of such an issue? I'm not entirely sure about what you're referring to

Reply:
Something along these lines:
```java
void f(@Unsafe Object unsafeValue) {
  @Safe Object result = switch (myEnum) {
      case ONE -> unsafeValue;
      case TWO -> "safe";
  }
}
```

Reply:
Thanks! Unfortunately, it seems like the following test is still failing. I'll add to a separate PR listing edge cases I've found and look deeper into it
```java
  import com.palantir.logsafe.*;
  import java.util.function.*;

  enum MyEnum {
    ONE,
    TWO
  }

  class Test {
    void f(MyEnum myEnum, @Unsafe Object unsafeValue) {
      @Safe
      Object result =
          switch (myEnum) {
              // BUG: Diagnostic contains: Dangerous
            case ONE -> unsafeValue;
            case TWO -> "safe";
          };
    }
  }
```